go/pb/lmsysmd/load/data/v1/data.go

```diff
@@ -0,0 +1,90 @@
+package data
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"regexp"
+	"sync"
+
+	"connectrpc.com/connect"
+	datav1 "github.com/Lev1ty/lmsysmd/pbi/lmsysmd/load/data/v1"
+	"github.com/Lev1ty/lmsysmd/pbi/lmsysmd/load/data/v1/datav1connect"
+	"github.com/jackc/pgx/v5/pgxpool"
+
+	"golang.org/x/oauth2/google"
+	"golang.org/x/oauth2/jwt"
+	"google.golang.org/api/drive/v3"
+	"google.golang.org/api/option"
+	"google.golang.org/api/sheets/v4"
+)
+
+type DataService struct {
+	datav1connect.UnimplementedDataServiceHandler
+	dbOnce sync.Once
+	db     *pgxpool.Pool
+	gOnce  sync.Once
+	gsSrv  *sheets.Service
+	gdSrv  *drive.Service
+	gConf  *jwt.Config
+}
+
+var re = regexp.MustCompile(`https:\/\/docs\.google\.com\/spreadsheets\/d\/([a-zA-Z0-9_-]+)\/`)
+
+var caseCategoryEnumMap = map[string]datav1.Data_CaseCategory{
+	"BR": datav1.Data_CASE_CATEGORY_BREAST,
+	"CH": datav1.Data_CASE_CATEGORY_CHEST,
+	"CV": datav1.Data_CASE_CATEGORY_CARDIOVASCULAR,
+	"GI": datav1.Data_CASE_CATEGORY_GASTROINTESTINAL,
+	"GU": datav1.Data_CASE_CATEGORY_GENITOURINARY,
+	"HN": datav1.Data_CASE_CATEGORY_HEAD_AND_NECK,
+	"MK": datav1.Data_CASE_CATEGORY_MUSCULOSKELETAL,
+	"NR": datav1.Data_CASE_CATEGORY_NEURORADIOLOGY,
+	"OB": datav1.Data_CASE_CATEGORY_OBSTETRIC,
+	"PD": datav1.Data_CASE_CATEGORY_PEDIATRIC,
+}
+
+var caseInputTypeEnumMap = map[string]datav1.Data_CaseInputType{
+	"HIST": datav1.Data_CASE_INPUT_TYPE_HISTORY,
+	"FIND": datav1.Data_CASE_INPUT_TYPE_FINDINGS,
+	"HXF":  datav1.Data_CASE_INPUT_TYPE_HISTORY_AND_FINDINGS,
+}
+
+func (ds *DataService) BatchCreateData(
+	ctx context.Context,
+	req *connect.Request[datav1.BatchCreateDataRequest],
+) (*connect.Response[datav1.BatchCreateDataResponse], error) {
+	ds.gOnce.Do(func() {
+		ds.gConf = &jwt.Config{
+			Email:      os.Getenv("GOOGLE_SERVICE_ACCOUNT"),
+			PrivateKey: []byte(os.Getenv("GOOGLE_SERVICE_ACCOUNT_PRIVATE_KEY")),
+			Scopes: []string{
+				"https://www.googleapis.com/auth/spreadsheets.readonly",
+				"https://www.googleapis.com/auth/drive.readonly",
+			},
+			TokenURL: google.JWTTokenURL,
+		}
+		var err error
+		if ds.gsSrv, err = sheets.NewService(ctx, option.WithHTTPClient(ds.gConf.Client(ctx))); err != nil {
+			panic(err)
+		}
+		if ds.gdSrv, err = drive.NewService(ctx, option.WithHTTPClient(ds.gConf.Client(ctx))); err != nil {
+			panic(err)
+		}
+	})
+	match := re.FindStringSubmatch(req.Msg.GetGoogleSheetUrl())
+	var spreadsheetId string
+	if len(match) > 1 {
+		spreadsheetId = match[1]
+	} else {
+		return nil, connect.NewError(connect.CodeInternal, fmt.Errorf("get spreadsheet id: could not parse spreadsheet id from url"))
+	}
+	_, err := ds.gsSrv.Spreadsheets.Values.Get(spreadsheetId, "Sheet1!A2:K").Do()
```

Comment:
should we avoid hardcoding in the range?

Reply:
The API requires you to specify a range to read from. This should be fine since we constrain the lab team to make the csv a specific number of columns anyways.

Reply:
Need to consider how it should look when/if we start onboarding other labs to the platform. Do we enforce a strict format of the sheet? If so, we should add validation to check if what they upload conforms to what is expected

Reply:
The plan is to enforce a string format of the sheet. We could implement a check on the column headers to see if they match the expected values. I don't think there is a way to check otherwise if the sheet is formatted correctly however the proto validation should prevent any badly formatted data from being introduced to the database.